src/mds/CInode.cc

```diff
@@ -3616,7 +3629,25 @@ int CInode::encode_inodestat(bufferlist& bl, Session *session,
   if (session->connection->has_feature(CEPH_FEATURE_DIRLAYOUTHASH)) {
     encode(file_i->dir_layout, bl);
   }
-  encode(xbl, bl);
```

Comment:
I don't really like c82b52191f81ac55353fa8f678cbff2ae64183fe. You're still copying the xattrs into a bufferlist (below). I think you could have simply done here:

`encode(std::move(xbl), bl)`

and then in `src/include/encoding.h`, add `inline void encode(buffer::ptr&& bp, bufferlist& bl)` which does `bl.claim_append(bl, 0)`.

Reply:
'bufferlist xbl; encode(xattrs, xbl);' does a 4k memory allocation.  It has nothing do with claim_append

Reply:
If you're saying `encode(xattrs,xbl)` does a 4k allocation, I'll take your word for it. But don't we also have that allocation here:

https://github.com/ceph/ceph/commit/c82b52191f81ac55353fa8f678cbff2ae64183fe#diff-1ddb64648a48a905cedbd7fb5e9359e0R3641

In which case are avoiding the allocation?

Reply:
we always needs to do memory allocation for MClientReply::trace_bl.  The new code encodes xattrs into trace_bl directly. trace_bl should have enough preallocated space in most cases. that's why this commit can avoids a memory allocation